File: src/sdk/typescript/src/utils/pool.ts
/**
 * Connection pool for managing HTTP connections
 */

import axios, { AxiosInstance } from 'axios';

export interface PoolConfig {
  maxConnections: number;
  maxIdleTime: number;
}

/**
 * Connection pool entry
 */
interface PoolEntry {
  client: AxiosInstance;
  lastUsed: number;
  inUse: boolean;
}

/**
 * Connection pool
 */
export class ConnectionPool {
  private pool: PoolEntry[] = [];
  private config: PoolConfig;
  private baseUrl: string;
  private headers: Record<string, string>;
  private cleanupInterval: NodeJS.Timer | null = null;

  constructor(
    baseUrl: string,
    headers: Record<string, string>,
    config: PoolConfig
  ) {
    this.baseUrl = baseUrl;
    this.headers = headers;
    this.config = config;
    this.startCleanup();
  }

  /**
   * Get a connection from the pool
   */
  async acquire(): Promise<AxiosInstance> {
    // Try to find an idle connection
    const idleConnection = this.pool.find(entry => !entry.inUse);

    if (idleConnection) {
      idleConnection.inUse = true;
      idleConnection.lastUsed = Date.now();
      return idleConnection.client;
    }

    // Create new connection if under limit
    if (this.pool.length < this.config.maxConnections) {
      const client = this.createClient();
      const entry: PoolEntry = {
        client,
        lastUsed: Date.now(),
        inUse: true
      };
      this.pool.push(entry);
      return client;
    }

    // Wait for a connection to become available
    return new Promise((resolve) => {
      const checkInterval = setInterval(() => {
        const available = this.pool.find(entry => !entry.inUse);
        if (available) {
          clearInterval(checkInterval);
          available.inUse = true;
          available.lastUsed = Date.now();
          resolve(available.client);
        }
      }, 10);
    });
  }

  /**
   * Release a connection back to the pool
   */
  release(client: AxiosInstance): void {
    const entry = this.pool.find(entry => entry.client === client);
    if (entry) {
      entry.inUse = false;
      entry.lastUsed = Date.now();
    }
  }

  /**
   * Create a new axios client
   */
  private createClient(): AxiosInstance {
    return axios.create({
      baseURL: this.baseUrl,
      headers: this.headers,
      timeout: 30000,
      validateStatus: () => true // Don't throw on any status
    });
  }

  /**
   * Start cleanup interval to remove idle connections
   */
  private startCleanup(): void {
    this.cleanupInterval = setInterval(() => {
      const now = Date.now();
      this.pool = this.pool.filter(entry => {
        if (entry.inUse) return true;
        if (now - entry.lastUsed > this.config.maxIdleTime) {
          return false; // Remove idle connection
        }
        return true;
      });
    }, this.config.maxIdleTime);
  }

  /**
   * Drain the pool and cleanup resources
   */
  async drain(): Promise<void> {
    if (this.cleanupInterval) {
      clearInterval(this.cleanupInterval);
    }

    this.pool.forEach(entry => {
      try {
        entry.client.defaults.timeout = 0;
      } catch {
        // Ignore cleanup errors
      }
    });

    this.pool = [];
  }

  /**
   * Get pool statistics
   */
  getStats(): { total: number; inUse: number; idle: number } {
    const inUse = this.pool.filter(e => e.inUse).length;
    return {
      total: this.pool.length,
      inUse,
      idle: this.pool.length - inUse
    };
  }
}

File: src/sdk/python/ai_agent/pool.py
"""
Connection pool for managing HTTP connections
"""

import asyncio
import logging
from typing import Optional
import aiohttp

logger = logging.getLogger(__name__)


class PoolConfig:
    """Connection pool configuration"""

    def __init__(self, max_connections: int = 10, max_idle_time: int = 60000):
        self.max_connections = max_connections
        self.max_idle_time = max_idle_time


class ConnectionPool:
    """Async connection pool"""

    def __init__(
        self,
        base_url: str,
        headers: dict,
        config: PoolConfig,
    ):
        self.base_url = base_url
        self.headers = headers
        self.config = config
        self.connector = aiohttp.TCPConnector(limit=config.max_connections)
        self.session: Optional[aiohttp.ClientSession] = None
        self._closed = False

    async def initialize(self) -> None:
        """Initialize the pool"""
        if self.session is None:
            self.session = aiohttp.ClientSession(
                connector=self.connector, headers=self.headers
            )
            logger.debug("Connection pool initialized")

    async def acquire(self) -> aiohttp.ClientSession:
        """Get a session from the pool"""
        if self.session is None:
            await self.initialize()
        return self.session

    async def close(self) -> None:
        """Close the pool"""
        if self.session:
            await self.session.close()
            self._closed = True
            logger.debug("Connection pool closed")

    def get_stats(self) -> dict:
        """Get pool statistics"""
        return {
            "connector_limit": self.connector.limit,
            "connector_limited_semaphore": self.connector._limit_sem._value
            if self.connector._limit_sem
            else None,
        }
